lightning/src/ln/quiescence_tests.rs

```diff
@@ -284,3 +285,128 @@ fn test_quiescence_tracks_monitor_update_in_progress_and_waits_for_async_signer(
+		check_added_monitors(&nodes[1], 1);
+	}
+	assert!(nodes[1].node.get_and_clear_pending_msg_events().is_empty());
+
+	// Finish the quiescence handshake.
+	nodes[0].node.handle_stfu(node_id_1, &stfu);
+	let stfu = get_event_msg!(&nodes[0], MessageSendEvent::SendStfu, node_id_1);
+	nodes[1].node.handle_stfu(node_id_0, &stfu);
+
+	nodes[0].node.exit_quiescence(&node_id_1, &chan_id).unwrap();
+	nodes[1].node.exit_quiescence(&node_id_0, &chan_id).unwrap();
+
+	// Now that quiescence is over, nodes are allowed to make updates again. nodes[1] will have its
+	// outbound HTLC finally go out, along with the fail/claim of nodes[0]'s payment.
+	let update = get_htlc_update_msgs!(&nodes[1], node_id_0);
+	check_added_monitors(&nodes[1], 1);
+	nodes[0].node.handle_update_add_htlc(node_id_1, &update.update_add_htlcs[0]);
+	if fail_htlc {
+		nodes[0].node.handle_update_fail_htlc(node_id_1, &update.update_fail_htlcs[0]);
+	} else {
+		nodes[0].node.handle_update_fulfill_htlc(node_id_1, &update.update_fulfill_htlcs[0]);
+	}
+	commitment_signed_dance!(&nodes[0], &nodes[1], update.commitment_signed, false);
+
+	if !fail_htlc {
+		expect_payment_claimed!(nodes[1], payment_hash2, payment_amount);
+	}
+
+	let events = nodes[0].node.get_and_clear_pending_events();
+	assert_eq!(events.len(), 3);
+	assert!(events.iter().find(|e| matches!(e, Event::PendingHTLCsForwardable { .. })).is_some());
+	if fail_htlc {
+		assert!(events.iter().find(|e| matches!(e, Event::PaymentFailed { .. })).is_some());
+		assert!(events.iter().find(|e| matches!(e, Event::PaymentPathFailed { .. })).is_some());
+	} else {
+		assert!(events.iter().find(|e| matches!(e, Event::PaymentSent { .. })).is_some());
+		assert!(events.iter().find(|e| matches!(e, Event::PaymentPathSuccessful { .. })).is_some());
+		check_added_monitors(&nodes[0], 1);
+	}
+	nodes[0].node.process_pending_htlc_forwards();
+	expect_payment_claimable!(nodes[0], payment_hash1, payment_secret1, payment_amount);
+
+	if fail_htlc {
+		nodes[0].node.fail_htlc_backwards(&payment_hash1);
+		let failed_payment = HTLCDestination::FailedPayment { payment_hash: payment_hash1 };
+		expect_pending_htlcs_forwardable_and_htlc_handling_failed!(&nodes[0], vec![failed_payment]);
+	} else {
+		nodes[0].node.claim_funds(payment_preimage1);
+	}
+	check_added_monitors(&nodes[0], 1);
+
+	let update = get_htlc_update_msgs!(&nodes[0], node_id_1);
+	if fail_htlc {
+		nodes[1].node.handle_update_fail_htlc(node_id_0, &update.update_fail_htlcs[0]);
+	} else {
+		nodes[1].node.handle_update_fulfill_htlc(node_id_0, &update.update_fulfill_htlcs[0]);
+	}
+	commitment_signed_dance!(&nodes[1], &nodes[0], update.commitment_signed, false);
+
+	if fail_htlc {
+		let conditions = PaymentFailedConditions::new();
+		expect_payment_failed_conditions(&nodes[1], payment_hash1, true, conditions);
+	} else {
+		expect_payment_claimed!(nodes[0], payment_hash1, payment_amount);
+		expect_payment_sent(&nodes[1], payment_preimage1, None, true, true);
+	}
+}
```

Comment:
Maybe similarly add comments thoughout here? Would be easier to follow which payment each block corresponds to.